_drafts/swim.md

```diff
@@ -0,0 +1,441 @@
+**I**nfection-style Process Group **M**embership Protocol — is the separation of
+the failure detection and membership update dissemination functionalities.
+
+### Failure detection component
+
+As the name suggests, the role of this component is to detect potential failures
+of nodes in the cluster.
+
+Like in heartbeating protocols, each node has a list of the nodes that are in
+the cluster. But instead of brutally pinging all of them every <code>T</code>
+interval of time, only one of them will be checked.
+
+On this schema, we see a node **randomly choosing** another one of its peers to
+check if it is still alive. It sends a <code>PING</code> message and expects an
+<code>ACK</code> in response, indicating that the target node is healthy.
+
+<figure>
+    <img src="/img/swim_ping_ack.svg" />
+    <figcaption>SWIM: node successfully pinged</figcaption>
+</figure>
+
+**If no <code>ACK</code>** from the node *C* **is received** by *A* in the pre-specified
+timeout period (determined by the message round-trip time, which is chosen
+smaller than the protocol period), then **an indirect check is performed**.
+
+*A* will select <code>k</code> members at random from its list of nodes and send
+them a <code>REQ-PING(C)</code> message. Each of these nodes will in turn send
+a <code>PING</code> message to *C* and forward the <code>ACK</code> (if
+received) to *A*.
+
+<figure>
+    <img src="/img/swim_req_ping.svg" />
+    <figcaption>SWIM: indirect checks from A to C</figcaption>
+</figure>
+
+At the end of the protocol period <code>T</code>, *A* checks if it has received
+any direct or indirect response from *C*. If not, it declares *C* as dead in its
+local membership list and hands this update off to the dissemination component.
+
+Indirect probing is a great way to avoid any congestion network path between
+two nodes and dropped packets on the, which might have caused the direct
+<code>PING</code> to have failed in the first place. That's also why the
+<code>ACK</code> is relayed back instead of being sent directly to the probing
+node. The number of false-positives decreases and the accuracy is improved.
+
+**Note:** the values <code>T</code>, <code>k</code> as well as the timeout
+period are parameters of the protocol and can be tuned to suit different needs
+and networks.
+
+### Dissemination component
+
+Upon detecting the failure of another group member, the process multicasts this
+information to the rest of the cluster as <code>failed(C)</code> message.
+A member receiving this message deletes the offending node from its local
+membership list. Information about newly joined members or voluntarily
+leaving members are disseminated in a similar manner.
+
+<figure>
+    <img src="/img/swim_multicast_failure.svg" />
+    <figcaption>SWIM: failure of C, detected and multicast by A</figcaption>
+</figure>
+
+**Note:** for a new node to join the cluster, it must know the address of
```

Comment:
I would have explained a bit more on the "automatic" discovery to join the cluster